content/articles/2019-02-purrr-0-3-0.Rmd

```diff
@@ -0,0 +1,308 @@
+
+
+## Stricter predicate checking
+
+purrr now checks the results of your predicate functions, which must now consistently return `TRUE` or `FALSE`. We no longer offer support for `NA` or for boolish numeric values (R normally interprets 0 as `FALSE` and all other values as `TRUE`). The purpose of this change is to detect errors earlier with a more relevant error messages.
+
+```{r, error = TRUE}
+keep(c(1, NA, 3), ~ . %% 2 == 0)
+```
+
+
+## Early termination of reduction
+
+`reduce()` and `accumulate()` now support early termination of the reduction. To halt the computation, just return the last value wrapped in a `done()` box:
+
+```{r}
+# This computes the total sum of the input vector
+reduce(1:100, ~ .x + .y)
+
+# This stops as soon as the sum is greater than 50
+reduce(1:100, ~ if (.x > 50) done(.x) else .x + .y)
+```
+
+This feature takes inspiration from the [Clojure](https://clojuredocs.org/clojure.core/reduced) language.
+
+
+## Rates
+
+Thanks to Richie Cotton (@richierocks) and Ian Lyttle (@ijlyttle), purrr gains a function operator to make a function call itself repeatedly when an error occurs.
+
+```{r, error = TRUE}
+counter <- 0
+
+f <- function(...) {
+  if (counter < 2) {
+    counter <<- counter + 1
+    stop("tilt!")
+  }
+  "result"
+}
+
+f()
+```
+
+If the function is wrapped with `insistently()`, it will try a few times before giving up:
+
+```{r}
+# Reset counter
+counter <- 0
+
+f2 <- insistently(f)
+f2()
+```
+
+Another rate limiting function is `slowly()`. While `insistently()` loops by itself, `slowly()` is designed to be used in your own loops, for instance in a map iteration:
+
+```{r}
+f <- function(...) print(Sys.time())
+
+walk(1:3, f)
+
+walk(1:3, slowly(f))
+```
+
+`slowly()` uses a constant rate by default while `insistently()` uses a backoff rate. The rate limiting can be configured with optional jitter via `rate_backoff()` and `rate_delay()`, which implement exponential backoff rate and constant rate respectively.
+
+```{r}
+walk(1:3, slowly(f, rate_backoff(max_times = Inf)))
+```
+
+
+## `map_if()`... or else?
+
+If you like using `map_if()`, perhaps you'll find the new `.else` argument useful. `.else` is a function applied to elements for which the predicate is `FALSE`:
+
+```{r}
+map_if(iris, is.numeric, mean, .else = nlevels)
+```
+
+
+## New `map_at()` features
+
+Colin Fay (@ColinFay) has added support for tidyselect expressions to `map_at()` and other `_at` mappers. This brings the interface of these functions closer to scoped functions from the dplyr package, such as `dplyr::mutate_at()`. Note that `vars()` is currently not reexported from purrr, so you need to use `dplyr::vars()` or `ggplot2::vars()` for the time being.
+
+```{r}
+suppressMessages(library("dplyr"))
+
+x <- list(
+  foo = 1:5,
+  bar = 6:10,
+  baz = 11:15
+)
+
+map_at(x, vars(starts_with("b")), mean)
+```
+
+`map_at()` now also supports negative selections:
+
+```{r}
+map_at(x, -2, `*`, 1000)
+```
+
+
+## Consistency of behaviour
```

Comment:
I think you could make this heading just "Consistency", and then "Behaviour" could be a level 3 header.